### PR TITLE
App Engine and Firebase Terraform

### LINKED ISSUE #1555: App Engine and Firebase Terraform
This is to the first set of changes to cover creation of the App Engine and Firebase projects using Terraform.

Philosophically, we want to use Infrastructure as code. Beyond normal IaC benefits, it specifically helps this project in duplicating the same set up between the dev and prod projects. Provides greater transparency and reproducibility.

## Checklist:
- [x] Searched the [existing issues](https://github.com/WorldHealthOrganization/app/issues) to ensure you are not creating a duplicate.
- [x] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md).
- [x] Issue has descriptive title focused on the feature or problem from the user's perspective

### PR DESCRIPTION
**Pull request by yukuairoy**
Closes #1555 

How did you test the change?

- [ ] iOS Simulator
- [ ] iOS Device
- [ ] Android Simulator
- [ ] Android Device
- [ ] `curl` to a dev App Engine server
- [ ] other, please describe

## Checklist:

- [X] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md) and verified that all contributions are properly licensed pursuant to the [LICENSE](https://github.com/WorldHealthOrganization/app/blob/master/LICENSE).

### CODE REVIEW

File: terraform/developer-guide.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  ```sh
+  export TF_VAR_billing_account=<GCP billing account id>
+  ```
+
+### Create or update a GCP project
+
+- Configure environment:
+
+  ```sh
+  export TF_VAR_project_name=<project name>
```
Comment (brunobowden) [inline terraform/developer-guide.md:24]: I think the instances should be specified in tf-var files. This puts it under source control: https://www.terraform.io/docs/configuration/variables.html#variable-definitions-tfvars-files

So the documented approach ends up with distinct files for prod, staging, hack, dev and so on. Staging and hack should be automatic pushes. Prod and dev can be manually done. Not sure if you removed `=` then the autocomplete would work which would be nice.

I like this better as the action is more obvious and not hidden by environment variables:
```
terraform apply -var-file=staging.tfvars
```

File: terraform/project.tf [hunk 2]
```diff
@@ -0,0 +1,31 @@
+variable "project_name" {}
```
Comment (brunobowden) [inline terraform/project.tf:1]: Repeat for all variables:
```suggestion
variable "project_name" {
  type = string
}
```

File: terraform/project.tf [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+variable "billing_account" {}
+variable "region" {}
+
+provider "google" {
+  region = var.region
+}
+
+resource "random_id" "id" {
+  byte_length = 4
+  prefix      = var.project_name
```
Comment (brunobowden) [inline terraform/project.tf:11]: Randomization is to make the name unique? I'd rather not have random fluff in the names as it's then much harder to remember and type. I think using the "who-myhealth2" as a prefix should suffice. You could declare standard shared config like this in a `terraform.tfvars` file which is automatically loaded anytime Terraform is run.

**Review by brunobowden — CHANGES_REQUESTED** [review]
@yukuairoy - this is great progress. I think we should move it to the `server/terraform` directory to keep the tree cleaner. 

Please also run `terraform fmt` just to check formatting. You can automatically enforce that later in #1620

**Comment by brunobowden** [comment]
@yukuairoy - you don't need to do everything here. Main thing, is to add something to description about testing:

Testing:
1. Deployed App Engine using modified script (I wonder if this should be defined within terraform or not?)
1. Tested App Engine instead with curl tests

File: terraform/project.tf [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+resource "google_project" "project" {
+  provider        = google-beta
+  name            = var.project_name
+  project_id      = random_id.id.hex
+  billing_account = var.billing_account
+  org_id          = 532343229286
```
Comment (epicfaace) [inline terraform/project.tf:19]: Would be best to make this a variable as well

File: terraform/project.tf [hunk 5]
```diff
@@ -0,0 +1,31 @@
+variable "project_name" {}
+variable "billing_account" {}
```
Comment (epicfaace) [inline terraform/project.tf:2]: It's also best practice to put all variables into a separate `vars.tf` file -- see https://github.com/cisagov/crossfeed/blob/master/infrastructure/vars.tf for an example.

File: terraform/developer-guide.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  export TF_VAR_region=<valid GCP region, e.g. europe-west6>
+  ```
+
+- Change directory into `terraform/` and do
+  ```sh
+  terraform init
+  ```
+- Deploy
+  ```sh
+  terraform apply
```
Comment (epicfaace) [inline terraform/developer-guide.md:34]: It's probably better practice to do the following (adding `out` to `.gitignore`, of course):

```
terraform plan -out out
terraform apply out
```

so the user is exactly sure about which changes will happen before they actually get applied

Comment (brunobowden) [reply terraform/developer-guide.md:34]: @epicfaace - Terraform by default shows you the changes plan before it's applied. I've used this many times and it's worked fine for me. The only thing that I think would be dangerous is `-auto-approve`:

```
terraform apply -auto-approve          Skip interactive approval of plan before applying.
```

**Comment by brunobowden** [comment]
@yukuairoy - I think it needs to specify `CLOUD_DATASTORE_COMPATIBILITY` to create the old Datastore database type upon initial creation: https://www.terraform.io/docs/providers/google/r/app_engine_application.html#database_type

File: server/terraform/project.tf [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+resource "google_firebase_project_location" "basic" {
+  provider    = google-beta
+  project     = google_firebase_project.default.project
+  location_id = var.region
+}
+
+resource "google_app_engine_application" "app" {
+  project       = google_firebase_project.default.project
+  location_id   = var.region
+  database_type = "CLOUD_FIRESTORE"
```
Comment (brunobowden) [inline server/terraform/project.tf:40]: Needs to be: `CLOUD_DATASTORE_COMPATIBILITY` until it's upgraded

File: server/terraform/developer-guide.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
+## How to create and setup GCP projects using terraform
+
+### Prerequisites
+
+- [Install Terraform](https://learn.hashicorp.com/tutorials/terraform/install-cli?in=terraform/gcp-get-started)
```
Comment (brunobowden) [inline server/terraform/developer-guide.md:5]: Update `server/README.md` to add the terraform install step. Devs will need this for server work to validate changes even if they don't edit the Terraform files.

File: server/terraform/project.tf [hunk 9]
```diff
@@ -0,0 +1,41 @@
+variable "project_name" {
+  type = string
+}
+variable "billing_account" {
+  type = string
+}
+variable "region" {
```
Comment (brunobowden) [inline server/terraform/project.tf:7]: Should this be `location_id` to match the convention for Google Cloud? Or is this a convention for Terraform?

File: server/terraform/project.tf [hunk 10]
```diff
@@ -0,0 +1,41 @@
+variable "project_name" {
```
Comment (brunobowden) [inline server/terraform/project.tf:1]: Should this be `project_id` to match the convention for Google Cloud? Or is this a convention for Terraform?

File: .github/workflows/policy-check.yaml [hunk 11]
```diff
@@ -39,3 +39,5 @@ jobs:
       # Match the tools/.lintstagedrc.json file
       - run: npx --no-install prettier -c '../**/*.{md,java,js,sh,yaml,yml}|*.{md,java,js,sh,yaml,yml}'
         working-directory: ./tools
+      - run: terraform fmt -check '../**/*.tf|*.tf'
```
[the remark this answers is not in the document]
Comment (brunobowden) [reply .github/workflows/policy-check.yaml:42]: Clearly this failed the CI. Try running this locally to see what works:
https://github.com/WorldHealthOrganization/app/pull/1619/checks?check_run_id=1283328236

File: tools/.lintstagedrc.json [hunk 12]
```diff
@@ -6,8 +6,8 @@
   "../**/*.dart|*.dart": [
     "flutter format"
   ],
-  "../**/*.tf|*.tf": [
-    "terraform fmt"
+  "../": [
```
Comment (brunobowden) [inline tools/.lintstagedrc.json:9]: I just ran it in the top level folder and it seemed to take a second to run. A little annoying to add that cost to everyone. I think this only runs it when there's a change that overlaps with the terraform directory:
```suggestion
  "../server/terraform/**": [
```

**Review by brunobowden — APPROVED** [review]
Just one minor issue and I think we're good to go. Feel free to defer the `terraform validate` to another PR.

File: server/terraform/defaults.auto.tfvars [hunk 13]
```diff
@@ -0,0 +1 @@
+location_id = "europe-west6"
```
Comment (brunobowden) [inline server/terraform/defaults.auto.tfvars:1]: Add comment:
```
# Switzerland is the preferred location for WHO
```

File: .github/workflows/policy-check.yaml [hunk 14]
```diff
@@ -39,3 +39,5 @@ jobs:
       # Match the tools/.lintstagedrc.json file
       - run: npx --no-install prettier -c '../**/*.{md,java,js,sh,yaml,yml}|*.{md,java,js,sh,yaml,yml}'
         working-directory: ./tools
+      - run: terraform fmt -check -recursive './terraform/'
+        working-directory: ./server
```
Comment (brunobowden) [inline .github/workflows/policy-check.yaml:43]: Don't split the working directory split across two lines:
```suggestion
      - run: terraform fmt -check -recursive .
        working-directory: ./server/terraform
```

File: tools/.lintstagedrc.json [hunk 15]
```diff
@@ -1,15 +1,9 @@
...
-  "../**/*.dart|*.dart": [
-    "flutter format"
-  ],
-  "../client/**": [
-    "./verify-client.sh"
-  ],
-  "../server/**": [
-    "./verify-server.sh"
-  ]
+  "../**/*.dart|*.dart": ["flutter format"],
```
Comment (brunobowden) [inline tools/.lintstagedrc.json:5]: Did you run prettier on this file? I'm fine with removing the empty line 2 but I think the rest of it works better with the original layout.

**Review by brunobowden — COMMENTED** [review]
@yukuairoy - please add some comments about testing as well.

**Comment by brunobowden** [comment]
@yukuairoy just told me over a phone call that this was good to merge in

**Pull request merged**